ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -1030,6 +1039,273 @@ private Pair<ServerErrorCode, GetResponse> getBlobFromRemoteReplica(ReplicateBlo
+      logger.trace("ReplicateBlobRequest replicated Blob {}, local Store has the Key", blobId);
+    } catch (StoreException e) {
+      // store.findKey throws e.getErrorCode() == StoreErrorCodes.ID_Not_Found if it doesn't exist.
+      if (e.getErrorCode() == StoreErrorCodes.ID_Not_Found) {
+        errorCode = ServerErrorCode.No_Error;
+      } else {
+        logger.error("ReplicateBlobRequest store.findKey throw exception {} on Blob {},", e.getErrorCode(), blobId);
+        errorCode = ErrorMapping.getStoreErrorMapping(e.getErrorCode());
+      }
+    } catch (Exception e) {
+      // getConvertedStoreKeys which may throw Exception
+      logger.error("ReplicateBlobRequest unknown exception to replicate {} of {}", blobId, replicateBlobRequest, e);
+      errorCode = ServerErrorCode.Unknown_Error;
+    }
+    if (errorCode != ServerErrorCode.No_Error) {
+      completeReplicateRequest(request, replicateBlobRequest, errorCode, startProcessTime);
+      return;
+    }
+
+    if (replicateBlobRequest.getOperationType() == RequestOrResponseType.TtlUpdateRequest) {
+      // Repair the PutBlob if local store doesn't have the blob when repair type is RequestOrResponseType.TtlUpdateRequest
+      if (localMessageInfo == null) {
+        Pair<ServerErrorCode, GetResponse> remoteGetResult = getBlobFromRemoteReplica(replicateBlobRequest);
+        errorCode = remoteGetResult.getFirst();
+        if (errorCode == ServerErrorCode.No_Error) {
+          errorCode = repairPutBlob(replicateBlobRequest, blobId, remoteGetResult.getSecond());
+        }
+      }
+      // now repair the TtlUpdate
+      if (errorCode == ServerErrorCode.No_Error) {
+        errorCode = repairTtlUpdate(replicateBlobRequest, blobId);
+      }
+    } else {
+      // replicateBlobRequest.getOperationType() == RequestOrResponseType.DeleteRequest
+      if (localMessageInfo != null) {
+        // local has the blob, we repair the delete record with information provided by the replicateBlobRequest
+        errorCode = repairDeleteRecordToLocalBlob(replicateBlobRequest, blobId);
+      } else {
+        // local doesn't have the blob,
+        // 1. try to replicate the blob from the source replica and repair the delete record.
+        // 2. if source doesn't have the PutBlob, get the tombstone from the source and repair the delete record.
+        Pair<ServerErrorCode, GetResponse> remoteGetResult = getBlobFromRemoteReplica(replicateBlobRequest);
+        errorCode = remoteGetResult.getFirst();
+        if (errorCode == ServerErrorCode.No_Error) {
+          errorCode = repairPutBlob(replicateBlobRequest, blobId, remoteGetResult.getSecond());
+          if (errorCode == ServerErrorCode.No_Error) {
+            errorCode = repairDeleteRecordToLocalBlob(replicateBlobRequest, blobId);
+          }
+        } else if (errorCode == ServerErrorCode.Blob_Deleted) {
+          errorCode = repairDeleteRecordFromSourceTombstone(replicateBlobRequest, blobId);
+        }
```

Comment:
I thought in V2, we don't use admin request to get lifeVersion and operation time from source anymore? I just force delete with lifeVersion being 0 operation time being now()?

Reply:
For V2, changed that we don't get the tombstone but just force delete with lifeVersion 0.